### DISCUSSION TITLE
How to transfer data from different material blocks?

### DISCUSSION THREAD

**Discussion by tqcthai** [idaholab/moose#25288] [Q&A General]
Dear all,

I am a new user in Moose.

I am implementing Moose for a mechanical coupling problem. In material block A, I compute the stress and a physical quality _a[_qp]. In material block B, which has type = ParsedMaterial, I need to specify the expression as a function of _a[_qp]. Do you know ho to do that in Moose? Is there any available example for that?

Thank you.
-T

**Comment by GiudGiud** [comment]
Hello

What is _a ? a function ? a material property ? A variable?
Once you have the name of a, you can use that name in the ParsedMaterial expression. 
You can look at the Parsedmaterial documentation for examples:
https://mooseframework.inl.gov/source/materials/ParsedMaterial.html

Guillaume

**Reply by tqcthai** [reply]
Hello GiudGiud,

_a[_qp] is an intermediated variable computed in the material block A, it is not a global variable specified in the input file.

**Reply by GiudGiud** [reply]
You ll have to declare a as a material property to be able to re-use it in another material property. 
Then in material A, you ll need to set _a

**Reply by tqcthai** [reply]
Yes, I declared it in Material A as _a(declareProperty<Real>(_base_name + "a"))
(I tried declareMaterialProperty instead of declareProperty but it was an error)
What is the syntax to set _a here in order to make it can be recognized in Material block B?

**Reply by GiudGiud** [reply]
in B, you can have `expression = '<base_name>a + ...'`
with `material_property_names = '<base_name>a'` as a parameter as well.

What is the base_name here? you ll need to use it in the expression in B

**Reply by tqcthai** [reply]
I also don't know what is _base_name here, it exists already in the example file, I removed it and tried as you suggested:

 In material A, cpp file: _a(declareProperty<Real>("a"))
and header file: MaterialProperty<Real> & _a


Material B block in the input file looks like:
```
type = DerivativeParsedMaterial
    property_name = block_B
    coupled_variables = 'c'
    material_property_names = 'a'
    expression = 'c^(2a)'
    derivative_order = 2
```
And I got the error: Cyclic dependency detected in object ordering

**Reply by GiudGiud** [reply]
does A depend on B? Or the material that computes A depend on B?
Because right now B depend on A

**Reply by tqcthai** [reply]
it should be block_B = c^(2a)

Where 'c' is a basis variable, and 'a' is computed from another basis variable in material A

**Reply by tqcthai** [reply]
it means 'a' does not depend on block_B or B.

**Reply by GiudGiud** [reply]
Is there any other material property computed at the same time as 'a' that does depend on B?

**Reply by tqcthai** [reply]
You are right, they are implicit dependent. But I am doing the transient analysis, let's say  block_B_{n+1} = c_{n+1}^(2a_{n}), now a is frozen as a material parameter of the previous step. I wonder how to do that in Moose.

**Reply by GiudGiud** [reply]
So you need to separate the calculations enough so that MOOSE does not see the dependency anymore (maybe add a material that ONLY computes a)
OR
retrieve the old material property in one of them (old = explicit, from the previous time step)
